### ISSUE TITLE
No error event bound for postgres dialect

### ISSUE THREAD

**Issue by myndzi** [tgriesser/knex#639]
I posted a comment on #595 but this deserves its own issue I think.

No error handler is bound to the pg connection object. You can see here:

https://github.com/tgriesser/knex/blob/master/lib/dialects/mysql/index.js#L107

MySQL has a "MySQL specific error handler" (what is MySQL specific about it?)

But...

https://github.com/tgriesser/knex/blob/master/lib/dialects/postgres/index.js#L106

No 'error' event is bound. It is, however, emitted in various cases; I presume the error in the callback is for errors opening the connection, but errors can still be emitted on a previously-opened connection (for example, when it is closed explicitly by the PG server). That's what I was working on testing when I ran into this. I was going to post another issue, but really this issue is the same problem, just under a different name.

**Comment by myndzi** [comment]
I didn't see the label get added, but I would think that synchronously crashing an application with nothing the user can do about it would get more attention that a 'code plz'.

The reason I didn't submit a patch is because I don't know the code well enough to know why the above behavior is as it is or how to correctly solve it, and I don't have an easy way to provide the tests for the patch.

Why does the MySQL instance say "mysql specific"? Should the problem be solved in a different way for pg? I need at least a comment on the problem from somebody who knows the codebase well.

**Comment by chanderson0** [comment]
I've been encountering this on Heroku, where my app will crash randomly (well, whenever Heroku or someone else trips the PG instances). I can reproduce it with `heroku pg:killall` - the app crashes, and doesn't seem to restart properly.

Copying the MySQL solution seems to mostly work for PG, so we have that up and running now.

https://github.com/chanderson0/knex/commit/1ba1c3996dd23875a64b473644193bb461babadd

The caveat is that the first request against the server after the error is passed up from PG will hang until it times out.

**Comment by bendrucker** [comment]
> the app crashes, and doesn't seem to restart properly

Let's first try to unpack that. Emitting an uncaught error event should cause the process to crash. If you've investigated why your application isn't restarting and reconnecting properly I'd like to hear about that. 

> The caveat is that the first request against the server after the error is passed up from PG will hang until it times out.

Hmm. So the connection should be destroyed but seems like that first request is still attempting to use it. First step here is getting tests against this behavior.

**Comment by myndzi** [comment]
Yeah, the problem is that the user of knex cannot catch the error. It is emitted from the PG library on an object that is held internally by knex.

**Comment by bendrucker** [comment]
Right, understood on the inability to fix this in userland (other than forking) right now. Will merge something right away if either of you manage to track down why that dead connection hangs around (or whatever else is causing the first query to fail).

**Comment by myndzi** [comment]
I was unable to reproduce the stalling, but the problem with your patch above is the "&& err.fatal" line -- this property is not set on the error that pg emits. Additionally, if (err) will fail when the function is triggered by the connection closing (without an error event), which is a problem that likely also affects mysql. The code should probably be simplified to:

    function connectionErrorHandler(client, connection, err) {
      if (!connection || connection.__knex__disposed) return;
      client.pool.genericPool.destroy(connection);
    }

With this and the event bindings added in acquireRawConnection, I receive no interruption to my queries, and no errors.

**Comment by myndzi** [comment]
See https://github.com/tgriesser/knex/pull/659

**Comment by chanderson0** [comment]
For what it's worth, I've been running the fixes from @myndzi in knex and generic-pool-redux for the last ~2 weeks and haven't encountered any issues so far.
